### PR TITLE
SchedulingExecutorServiceDelegate does not implement j.u.c.ScheduledExecutorService

### LINKED ISSUE #7611: ScheduledExecutorServiceDelegate violates contract of ScheduledExecutorService
Contract of `scheduleAtFixedRate` says: 
```
     * If any execution of this task
     * takes longer than its period, then subsequent executions
     * may start late, but **will not concurrently execute**.
```

However the `ScheduledExecutorServiceDelegate` wraps tasks in `ScheduledTaskRunner` which delegates to a different executor. As a consequence a task can be executed concurrently thus a contract of `ScheduledExecutorService` is violated. 

This leads to surprises. this was seen in a testing environment: 
```java
Exception in thread "cached4" java.lang.IllegalArgumentException: Comparison method violates its general contract!
    at java.util.TimSort.mergeHi(TimSort.java:895)
    at java.util.TimSort.mergeAt(TimSort.java:512)
    at java.util.TimSort.mergeCollapse(TimSort.java:437)
    at java.util.TimSort.sort(TimSort.java:241)
    at java.util.Arrays.sort(Arrays.java:1512)
    at java.util.ArrayList.sort(ArrayList.java:1454)
    at java.util.Collections.sort(Collections.java:175)
    at com.hazelcast.map.impl.eviction.ExpirationManager$ClearExpiredRecordsTask.sortPartitionContainers(ExpirationManager.java:127)
    at com.hazelcast.map.impl.eviction.ExpirationManager$ClearExpiredRecordsTask.run(ExpirationManager.java:121)
    at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:212)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
```

### PR DESCRIPTION
**Pull request by jerrinot**
Fixes #7611

Our scheduler has a different contract hence it should not
implement the interface from JDK.

I renamed it to the SchedulingExecutorServiceDelegate to avoid confusion.
Also I changed the method name from scheduleAtFixedRate to scheduleWithRepetition
to emphasize it has a different semantic.

The scheduleWithFixedDelay was removed as it never worked and it
effectively behaved as the scheduleAtFixedRate anyway.

### CODE REVIEW

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import com.hazelcast.core.ICompletableFuture;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.LoggingService;
+import com.hazelcast.spi.impl.executionservice.impl.SkipOnConcurrentExecutionDecorator;
```
Comment (jerrinot) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java]: I am not sure about this dependency here.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java [hunk 2]
```diff
@@ -149,19 +150,11 @@ public void run() {
...
-            }
-        }, initialDelay, period, unit);
-    }
-
-    @Override
-    public ScheduledFuture<?> scheduleWithFixedDelay(final Runnable command, long initialDelay, long period, TimeUnit unit) {
-        return scheduledExecutor.scheduleWithFixedDelay(new Runnable() {
-            public void run() {
-                executeInternalSafely(command, failureLoggingExecutionCallback);
+                executeInternalSafely(decoratedCommand, failureLoggingExecutionCallback);
```
Comment (jerrinot) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java]: here is a possible issue where the callback will be executed even when execution is skipped. I am not sure if that's a problem is reality.

**Comment by jerrinot** [comment]
It breaks SPI, but that's a good thing given the semantic shift.

**Comment by pveentjer** [comment]
I'm not too happy with the name SchedulingExecutorServiceDelegate since it interface (which is an abstraction of the implementation) clearly explains its implementation. Why not call it Scheduler?

File: hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java [hunk 3]
```diff
@@ -93,11 +93,11 @@ public LocalQueueStatsImpl createNew(String key) {
 
     public QueueService(NodeEngine nodeEngine) {
         this.nodeEngine = nodeEngine;
-        ScheduledExecutorService defaultScheduledExecutor
-                = nodeEngine.getExecutionService().getDefaultScheduledExecutor();
+        SchedulingExecutorService defaultSchedulingExecutorService
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: Default scheduler ----> I would rename it to the Global Scheduler. The one that is the working horse for all kinds of stuff.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SchedulingExecutorServiceDelegate.java [hunk 4]
```diff
@@ -27,12 +29,14 @@
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 
-final class ScheduledExecutorServiceDelegate implements ScheduledExecutorService {
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+public final class SchedulingExecutorServiceDelegate implements SchedulingExecutorService {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SchedulingExecutorServiceDelegate.java]: I would name this class SchedulingExecutorServiceImpl. It really isn't relevant that it is a delegate. Apart from that most services are called FooImpl.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SkipOnConcurrentExecutionDecorator.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Decorator to prevent concurrent task execution.
+ *
+ * If this task is running on a thread and another thread calls attempt the execute it concurrently
+ * then the 2nd execution will be skipped.
+ *
+ */
+public class SkipOnConcurrentExecutionDecorator implements Runnable {
+    private final AtomicBoolean isAlreadyRunning = new AtomicBoolean();
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SkipOnConcurrentExecutionDecorator.java]: Use a volatile int field and an AtomicIntegerFieldUpdater. This will reduce the amount of litter.

**Comment by pveentjer** [comment]
Looks good. Minor suggestions regarding naming.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SchedulingExecutorServiceDelegate.java (hunk 4 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SchedulingExecutorServiceDelegate.java]: well, it delegates to a given executor service. this is significant implementation aspect, the reason why we have it at all. otherwise we might very well use just the `ScheduledExecutorService` from JDK. 

what about `DelegatingSchedulingExecutorService` ?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SkipOnConcurrentExecutionDecorator.java (hunk 5 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SkipOnConcurrentExecutionDecorator.java]: `AtomicIntegerFieldUpdater` tends to be a tiny bit slower in current HotSpot. (I think the reason is that HotSpot does not fully trust final fields inside `Atomic*FieldUpdater`) 

I would have used the `Atomic*FieldUpdater` if the tasks were created with high rate. However here I assume a component schedules a task once only - during initialization/start. Hence the amount of objects created is not growing over time. 

If you still think it's good to use the Update I can change it, but I see little reasons to do so.

File: hazelcast/src/main/java/com/hazelcast/spi/SchedulingExecutorService.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ * guarantees a task won't be executed by multiple threads concurrently. The difference is that this service will simply
+ * skip a scheduled execution if another thread is still running the same task instead of postponing its execution.
+ * To emphasize this difference the method is called <code>scheduleWithRepetition</code>
+ * instead of <code>scheduleAtFixedRate</code>
+ *
+ * The other difference is this service does not offer an equivalent of
+ * {@link java.util.concurrent.ScheduledExecutorService#scheduleWithFixedDelay(Runnable, long, long, TimeUnit)}
+ *
+ */
+public interface SchedulingExecutorService extends ExecutorService {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/SchedulingExecutorService.java]: To me, it is hard to distinguish at first glance between names of `SchedulingExecutorService` and `ScheduledExecutorService`. (I don't have a better name now but want to state the problem)

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java (hunk 2 above)
Comment (ahmetmircik) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java]: Maybe you can add this potential problem as a note on that method.

**Comment by ahmetmircik** [comment]
:+1:  + minor comments

**Comment by sancar** [comment]
Since client does not have delegate executors for different logics, solution can be much simpler. I can convert internal executor to scheduled executor and give the same guarantees of scheduleWithFixDelay and scheduleWithFixRate.  @jerrinot I think this was also your first thought for server but could not do it because server needs `cachedExecutor`.

Client solution
https://github.com/hazelcast/hazelcast/pull/7646

In short, I am in favor of not touching to client in this pr and merge the #7646 because it will be much simpler.

**Comment by jerrinot** [comment]
@sancar: thanks for review! 

The issue with using the regular scheduler from JDK on the client side is that for example the QueryCache requires the a scheduler on both client and member. 

I believe we should offer a scheduler with the same semantic on both clients and members. 
When a client uses a regular JDK scheduler then it's hard to guarantee the same semantic. wdyt?

**Comment by sancar** [comment]
Hmm. QueryCache does not fit to my proposal. Then go ahead and merge this one. In #7646, I can still merge internal and scheduled executor to one scheduled executor for simplicity. But I will leave `scheduleWithRepetiton` in ClientExecutionService and that method will use `SkipOnConcurrentExecutionDecorator` to match the semantics.

**Comment by sancar** [comment]
@jerrinot  Is this going to be backported ?

**Comment by jerrinot** [comment]
@sancar: I think the backport is needed. at very least the behaviour-change should be backported as the current behaviour is causing some bad side-effects - see the original issue. 

I think I'll backport just the `SkipOnConcurrentExecutionDecorator` and use it in inside the `SchedulingExecutorServiceDelegate` and `ClientExService`, but I will not the the rename to maintain  API compatibility. WDYT?

**Comment by sancar** [comment]
@jerrinot Yes, that make sense. 
:+1:

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java (hunk 2 above)
Comment (jerrinot) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java]: I double checked it and it's not really an issue: the `ClientExecutionService` does not accept custom callback for `scheduleWithRepetition`. 

The service does use the callback internally, but only to logs errors -> not a problem.

File: hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java (hunk 3 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: done

File: hazelcast/src/main/java/com/hazelcast/spi/SchedulingExecutorService.java (hunk 6 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/SchedulingExecutorService.java]: I know. It's is a pickle. `ScheduledExecutorService` is a pretty good name for this, unfortunately Doug Lea already took it! :) 

I made the naming pattern consistent with `BufferingInputStream` which faced the same dilemma: similar functionality, slightly different contract.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SkipOnConcurrentExecutionDecorator.java (hunk 5 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/SkipOnConcurrentExecutionDecorator.java]: On this level I doubt if this tiny bit slower is going to be relevant.

I don't know how often these tasks are being created; if it really isn't often then leave it as it is.

File: hazelcast/src/main/java/com/hazelcast/spi/SchedulingExecutorService.java (hunk 6 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/SchedulingExecutorService.java]: I had to read multiple time to see the difference at all. So yes; please fix it. Else you also introduce a nasty class very everyone using Hazelcast since they will get this name in their popup.

I would go for 'Scheduler'

**Comment by pveentjer** [comment]
:+1: once you come up with a nice name for the scheduler

**Comment by jerrinot** [comment]
@ahmetmircik, @pveentjer: I changed the name to `TaskScheduler`. 

2 people raised the point the current name is too similar to the name from JDK. unambiguity has higher value then explicit indication it's kind of `ExecutorService`.

**Comment by pveentjer** [comment]
Exposing it as an ExecutorService could even be problematic. The ExecutorService exposes lifecycle methods like shutdown etc. These things you typically don't want to expose in an interface.

**Comment by pveentjer** [comment]
:+1:

**Comment by jerrinot** [comment]
@pveentjer: thanks for pointing it out. I understand methods like `shutdown()` should not be exposed to the user code. It should be managed by Hazelcast. I'll take a note.
